Add thread annotation macros

emersonknapp opened the pull request:
For clang thread safety analysis https://clang.llvm.org/docs/ThreadSafetyAnalysis.html

Related to https://github.com/ros2/ros2/issues/664

thomas-moulard commented on CHANGELOG.rst:
```diff
@@ -0,0 +1,7 @@
+
+0.2.0 (2019-02-25)
+------------------
+* Add clang thread safety analysis annotation macros
```
IIRC this one is generated by bloom, you don't need to write this manually (?)

emersonknapp replied:
Haha, ok, I was just trying to copy an existing package very closely, I will delete it

tfoote replied:
It need to be generated prior to bloom, there's catkin_generate_changelog which will fill the content in from the git history. We generally recommend manually revising it after the generation for better readabilty. And then catkin_prepare_release will add the version number and increment your package.xml too. https://index.ros.org/doc/ros2/Tutorials/Releasing-a-ROS-2-package-with-bloom/

thomas-moulard commented on package.xml:
```diff
@@ -2,8 +2,17 @@
 <package format="2">
   <name>rcpputils</name>
-  <version>0.1.0</version>
+  <version>0.2.0</version>
```
Any reason to start from 0.2.0?

emersonknapp replied:
Not really, I just committed the original empty package as 0.1.0

thomas-moulard commented:
Awesome - SGTM!

I think it's good we don't force everyone to compile with `-D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS` but we should also try not to forget to pass that in each package we convert as needed...

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6279)](http://ci.ros2.org/job/ci_linux/6279/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2701)](http://ci.ros2.org/job/ci_linux-aarch64/2701/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5158)](http://ci.ros2.org/job/ci_osx/5158/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6117)](http://ci.ros2.org/job/ci_windows/6117/)

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6284)](http://ci.ros2.org/job/ci_linux/6284/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2703)](http://ci.ros2.org/job/ci_linux-aarch64/2703/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5160)](http://ci.ros2.org/job/ci_osx/5160/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6119)](http://ci.ros2.org/job/ci_windows/6119/)

tfoote commented on package.xml:
```diff
@@ -6,4 +6,14 @@
   <description>Package containing various utility types and functions for C++</description>
```
It would be great to add some more to the description here to start a list of the things contained in this package. And also to have a link to documentation on how to use these macros.

emersonknapp replied:
Ok, is this the sort of thing we want in the package.xml, or in the README?

tfoote replied:
The README can and should be much more verbose. At the package.xml level it needs at least a high level summary of what's int the package. "various utilitity types and functions" doesn't tell you anything. This high level summary will be the first thing that people read, and will be redistributed into things like debian package descriptions.

wjwwood replied:
Personally, I never read this description, but giving some examples of what is in it would be useful.

For an example README, see `osrf_testing_tools_cpp` for what I consider to be a decent example:

https://github.com/osrf/osrf_testing_tools_cpp/blob/master/README.md

emersonknapp replied:
Have you seen the latest updated version? I added a README with information on usage - do you have specific comments on that?

wjwwood replied:
No, I think the current state is ok, just adding to what @tfoote said.

emersonknapp replied:
Ok, thanks!

emersonknapp commented:
I'll make a go at hitting all the macros in unit test. The first pass was mainly to make sure that including the header doesn't break the build

emersonknapp commented:
Updated the documentation, and updated tests to cover as many of the macros as possible, and left detailed analysis/documentation of the ones that could not be tested.

wjwwood commented on include/rcpputils/thread_safety_annotations.h:
```diff
@@ -0,0 +1,94 @@
+  return a;
+}
+
+#define RCPPUTILS_CAPABILITY(x) \
```
This is a bit bike-shed-y, but I think there will be lots of other utilities in the package at some point, so more namespacing than just `RCPPUTILS_` in these macros would be a good thing, imo. Maybe `RCPPUTILS_TSA_` (for Thread Safety Annotations) or `RCPPUTILS_TSAN_` (for Thread SANitizer).

Reading https://github.com/ros2/rmw_fastrtps/pull/259 I felt that things like `RCPPUTILS_REQUIRES` was pretty generic, and could use better namespacing.

emersonknapp replied:
Makes sense - changed it

emersonknapp commented:
Any feedback on latest revision?

emersonknapp commented:
Latest commit should fix CI issues - was able to reproduce locally by doing clang w/ libstdc++ - I had been only build gcc and clang+libcxx on the latest versions

tfoote commented:
Retriggered CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6407)](http://ci.ros2.org/job/ci_linux/6407/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2790)](http://ci.ros2.org/job/ci_linux-aarch64/2790/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5265)](http://ci.ros2.org/job/ci_osx/5265/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6239)](http://ci.ros2.org/job/ci_windows/6239/)

emersonknapp commented:
@tfoote with the build fix, are we ok to merge this?

wjwwood commented:
All commits have to be signed-off, the DCO check is currently failing.
